app/controllers/provider/sessions_controller.rb

```diff
@@ -75,6 +75,9 @@ def redirect_if_logged_in
 
   def authenticate_user
     strategy = Authentication::Strategy.build_provider(@provider)
+    captcha_is_available = request.post? # User & pass strategy
```

Comment:
just a suggestion - maybe we can move the `strategy = Authentication::Strategy.build_provider(@provider)` line to just before `strategy` is used?

Reply:
I think your quick fix is what makes best immediate sense.

My suggestion would be though to move this check to https://github.com/3scale/porta/blob/ad209e2004d7320daf28f8d751fb19bb32407bc7/app/lib/three_scale/bot_protection/controller.rb#L11

Because bot protection doesn't only mean captcha. Maybe we implement some other bot protection actions as well.

So for the separation of concerns principle, I think it makes more sense to move this bypass there.

And also add a TODO comment with what you have in PR description. That the class hierarchy needs to be fixed and some sort of a service check which one is the correct one.

Very nice investigation!

Reply:
> My suggestion would be though to move this check to
> 
> https://github.com/3scale/porta/blob/ad209e2004d7320daf28f8d751fb19bb32407bc7/app/lib/three_scale/bot_protection/controller.rb#L11
> 
> Because bot protection doesn't only mean captcha. Maybe we implement some other bot protection actions as well.
> 

What you say makes sense but it would imply knowing, from the bot protection module, which auth strategy is being used, and we can't know that until we refactor.

> And also add a TODO comment with what you have in PR description. 

Yeah, I'll do.